### PR TITLE
Fix dependency loading for real browsers

### PR DESCRIPTION
**Pull request by stefan-mihaila**
Fixes: https://github.com/resin-io-modules/resin-raven/issues/1

### CODE REVIEW

**Comment by stefan-mihaila** [comment]
@emirotin tell me how it looks to you. @jviotti suggested we should get rid of the dynamic requires all together, however, I have no idea of how to achieve that.

**Comment by jviotti** [comment]
@stefan-mihaila Maybe we can have two different entry points for each platform? For example, `require('resin-raven/browser')` and `require('resin-raven/node')` (potentially picking one as the default)?

**Comment by emirotin** [comment]
2 different entries would work for me, and if you also keep the `browser` field it will work automatically for the UI through webpack

File: src/services/external/node.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * This module handles dependency loading of the external services libraries in Node environments.
+ *
+ * @module node
+ */
+
+const isBrowserLike = require('./../is-browser-like');
```
Comment (emirotin) [inline src/services/external/node.js]: NIT: you don't need the leading dot, `require('../is-browser-like')` is fine

File: package.json [hunk 2]
```diff
@@ -2,7 +2,10 @@
   "name": "resin-raven",
   "version": "1.0.0-alpha.1",
   "description": "A wrapper around Sentry and Mixpanel that works in the browser, Node and Electron",
-  "main": "index.js",
+
+  "main": "src/services/external/node.js",
```
Comment (emirotin) [inline package.json]: I don't quite get this. These files don't export anything. They require the top-level file (index) and inject some of the dependencies, but don't return anything to the outer world.

So when I do `require('resin-raven')` I would get an empty exports object.

Am I wrong?

I would imagine them looking like

```
const resinRavenFactory = require('./common')

module.exports = resinRavenFactory({
  SentryLib: require('raven-js/dist/raven'),
  MixpanelLib: require('mixpanel-browser')
})
```

Also reaching out to the higher-level modules is usually a code smell, normally the parent should assmeble its children.

**Comment by emirotin** [comment]
I have a general question: why does this module include mixpanel? Shoudn't it be responsible for doing one thing - packaging Raven for cross-environment usage.

Also if it's still required can you make use of https://github.com/resin-io-modules/resin-analytics/tree/master/packages/resin-universal-mixpanel instead? It's already a Universal package, and it does browser-friendly asynchronous loading of mixpanel lib. You can see its usage examples here: https://github.com/resin-io-modules/resin-analytics/tree/master/packages/resin-mixpanel-client (which may be actually more suitable for your needs).

Overall the resin-analytics repo can be used as a reference point for dealing with universal modules.

**Comment by stefan-mihaila** [comment]
@emirotin The original idea was to have a wrapper around Raven, but then we decided to make it a wrapper around both Raven and Mixpanel; it's also supposed to change to not be just a general wrapper, but instead provide a lot of default configuration (f.e. I will look at what context is sent to external services by Etcher, Resin Dashboard, Resin Admin Panel and send what I can by default).

I didn't actually have any idea of `resin-universal-mixpanel` and `resin-mixpanel-client`.

CC: @jviotti

File: package.json (hunk 2 above)
Comment (stefan-mihaila) [reply package.json]: You're right, it did absolutely no sense what I did. Updated PR.

**Comment by stefan-mihaila** [comment]
Sorry for the stupid question, but I'm not sure what having two different entry points mean. Something like https://github.com/resin-io-modules/resin-raven/blob/53bda3f12f5a182b6201a9bac5106b9a87f80827/src/main.js#L19-L19? If so, isn't that still a dynamic require?

**Comment by jviotti** [comment]
@stefan-mihaila I mean creating two files: one that only requires the browser stuff, and and other one that requires the node stuff. Neither of those will have dynamic requires.

For example, you can place two files at the root of the project:

- `browser.js`
- `node.js`

Then, we can require the browser file by doing `require('resin-raven/browser')`, and `require('resin-raven/node')` for the node one.

**Comment by stefan-mihaila** [comment]
@jviotti That's what I did in this PR yesterday, I did create a `browser.js` and a `node.js`. But in order to pick one or the other, that's still a dynamic require, isn't it? Or is there a way to specify different entry points in `package.json`?

**Comment by stefan-mihaila** [comment]
@jviotti @emirotin I'm thinking of using `resin-universal-mixpanel` in another PR, don't want to bloat this one any more than it already is. I think if this fixes correctly https://github.com/resin-io-modules/resin-raven/issues/1 and if @jviotti is happy with the entry points, we can merge.

The PR using `resin-universal-mixpanel` should also change the API and return promises as `resin-universal-mixpanel` does.

What do you think?

**Comment by jviotti** [comment]
@stefan-mihaila 

> @jviotti That's what I did in this PR yesterday, I did create a browser.js and a node.js. But in order to pick one or the other, that's still a dynamic require, isn't it? Or is there a way to specify different entry points in package.json?

I don't think I follow. The user should know in advance which one it requires. In the case of Etcher, we will require the node one, while the UI will require the browser one. Am I missing something?

**Comment by stefan-mihaila** [comment]
@jviotti, it was me who was not understanding. I was thinking the lib should have an entry-point, so you can do `require('resin-raven')`. You are suggesting to do `require('resin-raven').node` (or `require('resin-raven/node')`. I'll update.

**Comment by jviotti** [comment]
@stefan-mihaila Not quite. Making the module expose two objects (`node` and `browser`), would require dynamic requires.

What I'm saying is that we can do `require('resin-raven/node')` or `require('resin-raven/browser')` which is different from `require('resin-raven').node` and `require('resin-raven').browser`.

In the first case, we have two unrelated files in the top level of the module, while in the latter case, a single entry point exports two object properties.

**Comment by stefan-mihaila** [comment]
@jviotti updated

**Review by jviotti — APPROVED** [review]

**Comment by emirotin** [comment]
Sorry for the delayed reply.
As I've mentioned somewhere properly declaring the two entry points (as done now, so all looks good from that point) will work this way:
- normal requires (like in node.js) will follow the `main` entry
- bundlers (packaging the code for the browser with Webpack or Browserify) will tolerate the `browser` entry and this skip requiring the node stuff altogether

**Comment by emirotin** [comment]
On a general note I don't think naming this `resin-raven` and bundling Mixpanel with it makes sense.
If that's what you need for etcher you can call the project `etcher-analytics` or `etcher-monitoring` or something like that. The thing is we already have the browser analytics in the UI with the help of `resin-analytics` module and we don't want another Mixpanel instance hanging around. 

While having streamlined cross-platform Sentry module makes total sense.

**Comment by jviotti** [comment]
> If that's what you need for etcher you can call the project etcher-analytics or etcher-monitoring or something like that. The thing is we already have the browser analytics in the UI with the help of resin-analytics module and we don't want another Mixpanel instance hanging around.

It sounds to me that `resin-analytics` should be re-used for in Etcher as well. If all projects (UI, CLI, Etcher) re-use both `resin-analytics` and `resin-raven`, then we could merge the two as a single module, that we all use, right?

**Comment by emirotin** [comment]
Yes sure. `resin-analytics` is the mono-repo where the 2 modules are just generic universal wrappers around mixpanel and GA (that use the friendly lazy-loading in the browser and the proper node libs in node), resin-mixpanel-client is the convenience wrapper (also universal, of course), the http mock lib is the testing helper (also generic), and finally the `resin-event-log` is the core project there that tracks the events.

So I think you can start using it in etcher and see how it goes, and keep this project for an easy to use one-stop universal Raven integration.

**Comment by stefan-mihaila** [comment]
@emirotin You're totally right about the naming, and `resin-analytics` is a much better name for its purpose. It would probably make sense for me to write a `Sentry` wrapper in `resin-analytics` and use `resin-analytics` and just drop this whole repo.

One thing that I don't like, is that the code style used in `resin-analytics` is different of what we use in `Etcher` and what I used in this repo. I really think we should have some coding standards company-wide. I kind of like https://github.com/airbnb/javascript which are pretty close to what we use in `Etcher`.

I wanted to get something running in `Etcher` asap, and my thought process was that making PRs in `resin-analytics` would slow me down. I'll talk with @jviotti about this in the `Etcher clusters` call and take some decisions.

Some things that I would need to change in `resin-analytics`:

- There was a small problem when I tried to use `mixpanel-universal-client` with `resin-raven`, as using `import('resin-raven/browser')` from the `Electron` renderer thread would actually use the Mixpanel node library instead of the web one. I decided to drop `mixpanel-universal-client` because I really wanted to get as quickly as possible something running in `Etcher`.

- in `resin-raven` we went for sending some default context: https://github.com/resin-io-modules/resin-raven/blob/13c73a760a15de7d2f5628b06b4b89ac0da7163e/src/default-context.js#L43-L43 and disabling logging when we're running in `Electron` but not in `asar`: https://github.com/resin-io-modules/resin-raven/blob/39ea2e94aeb028608a4a78afc1c70f027a72103c/browser.js#L25-L25. Probably it would make sense to do something similar in node, maybe depending on `NODE_ENV`, and maybe in the browser also.

- [`flattenStartCase`](https://github.com/resin-io-modules/resin-raven/blob/418313da1a874a68d140971dd0be70af305ac6e2/src/utils.js#L49-L49)

- [`hideAbsolutePathsInObject`](https://github.com/resin-io-modules/resin-raven/blob/418313da1a874a68d140971dd0be70af305ac6e2/src/utils.js#L117-L117)

Anyways, I'll sync with @jviotti in an hour and if needed we can all three have a short call.

**Comment by jviotti** [comment]
The conclusion was the following: lets rename this repo to something not Sentry specific, and make use of it on Etcher ASAP, since this is blocking the next release. Then, we can start back-porting stuff to `resin-analytics`, and eventually switch to it on Etcher.

**Comment by emirotin** [comment]
Totally agree on the last suggestion, Etcher shouldn't be blocked here.

> One thing that I don't like, is that the code style used in resin-analytics is different of what we use in Etcher

Well, obviously it didn't make much sense for me to look into the ethcer style before authoring that modules. I believe the code there is more or less standard ES6 (and in general code style bikeshedding is not very helpful :P)

> There was a small problem when I tried to use mixpanel-universal-client with resin-raven, as using import('resin-raven/browser') from the Electron renderer thread would actually use the Mixpanel node library instead of the web one. 

Of course, I have explained that above. The browser part kicks in only when you're using the **bundler** (like Webpack) that understands the `browser` field in package.json. That's how things work.

> I decided to drop mixpanel-universal-client because I really wanted to get as quickly as possible something running in Etcher.

Definitely makes sense short-term.

> Probably it would make sense to do something similar in node, maybe depending on NODE_ENV, and maybe in the browser also.

I think just make an option like `collectContext` and set it to true from Electron apps when it makes sense for you, but let other projects decide and pass the option.

`flattenStartCase, hideAbsolutePathsInObject` — didn't get these items (also please think up with @pimterry as he's currently working on the analytics-related things in the UI and the API).

**Comment by stefan-mihaila** [comment]
> (and in general code style bikeshedding is not very helpful :P)

Totally agree, it was more like a "can we do something about it so we don't have different code styles for every project we start?" rather than "my code style is better than yours". Sorry if it sounded that way.

`flattenStartCase` is there because Mixpanel wants flattened objects and we log nested objects in `Etcher`, while `hideAbsolutePathsInObject ` is there because we decided to send the `basename` instead of the full path (for files) to analytics services, as the full path contains sensitive information as the username. We wanted to make this project require as little configuration as possible from the user, rather than a wrapper that allows you to do anything.

@pimterry Take a look at `resin-corvus` in its current form and let me know if it's something you could use, or you need some other things from it. You can now pass `disableConsoleAlerts: true` to it: https://github.com/resin-io-modules/resin-corvus/blob/master/src/resin-corvus.js#L124

**Comment by pimterry** [comment]
@stefan-mihaila Thanks. With disableConsoleAlerts in there too, it looks like the CLI could be updated to use that pretty easily. I'm probably going to hold off until we've got this tested in Etcher and everything settles down into whatever form we're going with, but I think it'll be an easy change to make in the CLI when we do.

On resin-analytics, do watch out if you're changing or doing anything substantial on the mixpanel integration. I'm currently updating it in a few exciting ways and the api is going to change slightly very soon there to make this more consistent and reliable in the UI. Shouldn't be a problem outside mixpanel though, and it sounds like you're not integrating with resin-analytics for that yet, so hopefully this is fine, but just fyi!

**Comment by emirotin** [comment]
> Totally agree, it was more like a "can we do something about it so we don't have different code styles for every project we start?" rather than "my code style is better than yours". Sorry if it sounded that way.

No, it definitely didn't, just saying if the code style is sane and consistent it should be fine. Yes it would be nice to standardize things, but I don't think it's a valid reason not to use something as a dependency or to block such usage before the styles are consolidated.

**Pull request merged**